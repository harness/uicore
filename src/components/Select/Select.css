.main {
  position: relative;

  :global(.bp3-input) {
    display: block;
    border: 1px solid var(--grey-200);
    border-radius: var(--spacing-2);
    font-size: var(--font-size-normal);
    padding: var(--spacing-small);
    height: var(--spacing-8);
    color: var(--grey-900);
    box-shadow: none;
    overflow: hidden;
    text-overflow: ellipsis;
    word-break: break-all;

    &:hover {
      border-color: var(--primary-6);
      box-shadow: 0px 0px 1px rgba(40, 41, 61, 0.04), 0px 2px 4px rgba(96, 97, 112, 0.16);
    }

    &:focus {
      border-color: var(--primary-7);
    }

    &::-moz-placeholder,
    &::-webkit-input-placeholder,
    &:-ms-input-placeholder,
    &:-moz-placeholder,
    &::placeholder {
      color: var(--grey-400) !important;
      font-weight: 400;
    }

    &:disabled {
      background-color: var(--grey-50);
      color: var(--grey-400);
      border-color: var(--grey-200);
      box-shadow: none;
    }
  }

  :global(.bp3-input-group .bp3-input-left-container) {
    margin-top: 2px;
  }
  :global(.bp3-input-group.bp3-large .bp3-input-left-container) {
    margin-top: 0;
  }

  & :global(.bp3-overlay) {
    display: block;
    width: 100%;
  }

  :global {
    .bp3-input-action {
      opacity: 0.5;
    }
  }

  & :global(.bp3-transition-container),
  & :global(.bp3-popover) {
    width: 100%;
    box-shadow: none;
    border-radius: 0 0 4px 4px;
  }

  & :global(.bp3-menu) {
    max-height: 350px;
    width: 100%;
    overflow: auto;
    border-radius: var(--spacing-2);
    padding: 0;
    position: absolute;
    top: var(--spacing-1);
    border: 0.5px solid rgba(0, 0, 0, 0.1);
    box-sizing: border-box;
    box-shadow: 0px 0px 1px rgb(40 41 61 / 4%), 0px 2px 4px rgb(96 97 112 / 16%);
  }

  & .menuItem {
    height: var(--spacing-8);
    padding: 0 var(--spacing-small);
    display: flex;
    align-items: center;
    cursor: pointer;

    .menuItemLabel {
      color: var(--black);
      font-weight: 400;
      overflow: hidden;
      text-overflow: ellipsis;
      word-break: break-all;
    }

    :global(.bp3-icon) {
      margin-right: var(--spacing-small);
    }

    &.active {
      background: var(--primary-1);
    }

    &.menuItemSizeSmall {
      height: 24px;
      font-size: var(--font-size-small);
    }

    &.menuItemSizeLarge {
      height: 40px;
      font-size: var(--font-size-medium);
    }

    &.disabled {
      color: var(--grey-400);
      cursor: not-allowed;
    }

    &.loading {
      color: var(--grey-400);
      cursor: not-allowed;
    }

    &:hover {
      background: var(--primary-1);
    }
  }

  :global(.icon-container) {
    position: absolute;
    right: 0;
  }
}
