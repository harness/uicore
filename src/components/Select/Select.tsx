/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import cx from 'classnames'
import { Position, Classes } from '@blueprintjs/core'
import { Suggest, ISuggestProps, IItemRendererProps } from '@blueprintjs/select'

import css from './Select.css'
import { Button } from '../../components/Button/Button'
import { Icon, IconProps } from '@harness/icons'
import { Utils } from '../../core/Utils'
import { Text } from '../../components/Text/Text'
import { Popover } from '../../components/Popover/Popover'

export interface SelectOption {
  label: string
  value: string | number | symbol
  icon?: IconProps
}

export enum SelectSize {
  Small = 'small',
  Medium = 'medium',
  Large = 'large'
}

type Props = ISuggestProps<SelectOption>

const Loading = Symbol('loading')

export interface SelectProps
  extends Omit<
    Props,
    | 'popoverProps'
    | 'inputValueRenderer'
    | 'itemRenderer'
    | 'onItemSelect'
    | 'query'
    | 'selectedItem'
    | 'items'
    | 'activeItem'
    | 'onActiveItemChange'
  > {
  inputValueRender?: Props['inputValueRenderer']
  itemRenderer?: Props['itemRenderer']
  onChange?: Props['onItemSelect']
  value?: Props['selectedItem']
  size?: SelectSize
  items: Props['items'] | (() => Promise<Props['items']>)
  allowCreatingNewItems?: boolean
  name?: string
  whenPopoverClosed?: (node: HTMLElement) => void
  addClearBtn?: boolean
  usePortal?: boolean
  popoverClassName?: string
  onQueryChange?: Props['onQueryChange']
  addTooltip?: boolean
  borderless?: boolean
}

function getIconSizeFromSelect(size: SelectSize = SelectSize.Medium) {
  let iconSize = 16
  if (size === SelectSize.Small) {
    iconSize = 12
  } else if (size === SelectSize.Large) {
    iconSize = 18
  }
  return iconSize
}

export function defaultItemRenderer(
  item: SelectOption,
  props: IItemRendererProps,
  size: SelectSize = SelectSize.Medium
): JSX.Element | null {
  if (!props.modifiers.matchesPredicate) {
    return null
  }

  if (item.value === Loading) {
    return (
      <li key="loading" className={cx(css.menuItem, css.loading)}>
        Loading results...
      </li>
    )
  }

  return (
    <li
      key={item.value.toString()}
      className={cx(
        css.menuItem,
        {
          [css.active]: props.modifiers.active,
          [css.disabled]: props.modifiers.disabled
        },
        { [css.menuItemSizeSmall]: size === SelectSize.Small },
        { [css.menuItemSizeLarge]: size === SelectSize.Large }
      )}
      onClick={props.handleClick}>
      {item.icon ? <Icon size={getIconSizeFromSelect(size)} {...item.icon} /> : null}
      <Text className={css.menuItemLabel} lineClamp={1}>
        {item.label}
      </Text>
    </li>
  )
}

export function createNewItemFromQuery(query: string): SelectOption {
  return { label: query, value: query }
}

export function Select(props: SelectProps): React.ReactElement {
  const [query, setQuery] = React.useState('')
  const [loading, setLoading] = React.useState(false)
  const [items, setItems] = React.useState(Array.isArray(props.items) ? props.items : [])
  const {
    onChange,
    value,
    size,
    itemRenderer,
    addClearBtn,
    whenPopoverClosed,
    popoverClassName = '',
    resetOnSelect = true,
    resetOnClose = true,
    addTooltip = false,
    borderless = false,
    ...rest
  } = props
  const [item, setItem] = React.useState<SelectOption | undefined | null>(undefined)
  const showClearBtn =
    !!addClearBtn && value !== null && typeof value?.value !== undefined && value?.value !== null && value?.value !== ''

  React.useEffect(() => {
    setItem(value)
  }, [value])

  function handleItemSelect(item: SelectOption) {
    if (item.value === Loading) {
      return
    }
    if (typeof onChange === 'function') {
      onChange(item)
    } else {
      setItem(item)
    }
  }

  React.useEffect(() => {
    if (Array.isArray(props.items)) {
      setItems(props.items)
    } else if (typeof props.items === 'function') {
      setLoading(true)
      const promise = props.items()

      if (typeof promise.then === 'function') {
        promise.then(results => {
          setItems(results)
          setLoading(false)
        })
      } else {
        setLoading(false)
      }
    }
  }, [props.items])

  function createNewItemRenderer(
    query: string,
    _active: boolean,
    handleClick: React.MouseEventHandler<HTMLElement>
  ): JSX.Element | undefined {
    if (loading) {
      return (
        <li key="loading" className={cx(css.menuItem, css.loading)}>
          Loading results...
        </li>
      )
    }

    if (
      !loading &&
      items.filter(item => item.label.toString().toLowerCase().includes(query.toLowerCase())).length === 0
    )
      return (
        <React.Fragment>
          <div className={css.noResultsFound}>Nothing Found</div>
          {props.allowCreatingNewItems && (
            <Button
              intent="primary"
              minimal
              text={query}
              icon="plus"
              className={css.createNewItemButton}
              onClick={handleClick as React.MouseEventHandler<Element>}
            />
          )}
        </React.Fragment>
      )
  }
  const renderSuggestComponent = () => (
    <Suggest
      inputValueRenderer={opt => opt.label}
      itemRenderer={(item: SelectOption, props: IItemRendererProps) =>
        itemRenderer?.(item, props) || defaultItemRenderer(item, props, size)
      }
      itemListPredicate={(query, items) =>
        items.filter(item => item.label.toString().toLowerCase().includes(query.toLowerCase()))
      }
      createNewItemFromQuery={props.createNewItemFromQuery || createNewItemFromQuery}
      createNewItemRenderer={props.createNewItemRenderer || createNewItemRenderer}
      noResults={<NoMatch />}
      {...rest}
      inputProps={{
        onChange(e: React.ChangeEvent<HTMLInputElement>) {
          setQuery(e.target.value)
        },
        value: query,
        placeholder: Utils.getSelectComponentPlaceholder(rest?.inputProps?.placeholder),
        leftElement: item?.icon ? <Icon size={getIconSizeFromSelect(size)} {...item?.icon} /> : undefined,
        rightElement: (
          <>
            {!props.disabled && showClearBtn ? (
              <Icon
                name="main-delete"
                onClick={(e: React.MouseEvent<HTMLHeadingElement, MouseEvent>) => {
                  e.preventDefault()
                  handleItemSelect({ value: '', label: '' })
                }}
                size={14}
                padding={{ top: 'small', right: 'xsmall', bottom: 'small' }}
              />
            ) : null}
            <Icon
              name={'chevron-down'}
              onClick={e => {
                const input = e.currentTarget.parentElement?.previousElementSibling as HTMLInputElement
                input?.focus()
              }}
              size={14}
              padding={size === SelectSize.Small ? 'xsmall' : 'small'}
            />
          </>
        ),
        small: size === SelectSize.Small,
        large: size === SelectSize.Large,
        name: props.name,
        ...props.inputProps
      }}
      resetOnSelect={resetOnSelect}
      resetOnClose={resetOnClose}
      items={loading ? [{ label: 'Loading...', value: Loading }] : items}
      selectedItem={item}
      onItemSelect={handleItemSelect}
      query={query}
      popoverProps={{
        targetTagName: 'div',
        fill: true,
        usePortal: !!props.usePortal,
        minimal: true,
        position: Position.BOTTOM_LEFT,
        className: cx(css.main, { [css.borderless]: borderless }),
        popoverClassName: cx(css.popover, popoverClassName),
        onClosed: whenPopoverClosed,
        modifiers: {
          preventOverflow: {
            // This is required to always attach the portal to the start of the reference instead of the middle
            escapeWithReference: !!props.usePortal
          },
          offset: {
            // This is required to offset the portal after it is attached to the reference.
            // By default the portal is positioned at top: 0, left:0 wrt it's reference
            offset: props.usePortal ? '1 2' : 0
          }
        }
      }}
    />
  )
  const tooltipContent = item?.label ? (
    <div className={css.tooltipContainer} color="white">
      {item.label}
    </div>
  ) : (
    ''
  )
  return addTooltip ? (
    <Popover
      boundary="viewport"
      interactionKind="hover"
      content={tooltipContent}
      isDark={true}
      fill={true}
      popoverClassName={Classes.DARK}
      position="bottom">
      {renderSuggestComponent()}
    </Popover>
  ) : (
    renderSuggestComponent()
  )
}

export function NoMatch(): React.ReactElement {
  return <li className={cx(css.menuItem, css.disabled)}>No matching results found</li>
}
