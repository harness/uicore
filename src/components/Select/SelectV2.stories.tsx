/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React, { useState } from 'react'
import type { Meta, Story } from '@storybook/react'

import { SelectV2, SelectV2Props, Layout, Button } from '../..'
import { Title, Subtitle, ArgsTable, Stories, PRIMARY_STORY, Primary } from '@storybook/addon-docs/blocks'
import data from '../../_stories/components/pokedex.json'
import { omit } from 'lodash-es'
export default {
  title: 'Form / SelectV2',

  component: SelectV2,
  parameters: {
    docs: {
      source: {
        type: 'code'
      },

      page: function PageDescription() {
        return (
          <>
            <Title>SelectV2</Title>

            <Subtitle>
              <pre>
                <code>{`import { SelectV2 }  from '@harness/uicore'`}</code>
              </pre>
            </Subtitle>
            <Primary />
            <ArgsTable story={PRIMARY_STORY} />

            <Stories />
          </>
        )
      }
    }
  },
  decorators: [Story => <Story />]
} as Meta

interface SelectOption {
  label: string
  value: string | number | symbol
  icon?: any
}

const itemsLocal: SelectOption[] = data.map(row => ({
  label: row.name,
  value: row.id,
  icon: row.icon
}))

function dummyPromise(): Promise<SelectOption[]> {
  return new Promise<SelectOption[]>(resolve => {
    setTimeout(() => {
      resolve(itemsLocal)
    }, 5000)
  })
}

export const SimpleSelect: Story<SelectV2Props> = args => {
  const [val, setVal] = useState<SelectOption>()
  const { items = itemsLocal } = args
  const argsCopy = omit(args, ['items', 'itemRenderer', 'tagRenderer'])
  return (
    <Layout.Horizontal id="primary-borderless-buttons">
      <SelectV2
        items={items}
        value={val}
        onChange={setVal}
        itemRenderer={(item, props) => (
          <Button
            style={{ width: '100%', display: 'block' }}
            minimal
            onClick={ev => {
              props.handleClick(ev as any)
            }}>
            {item.label}
          </Button>
        )}
        {...argsCopy}>
        <Button
          intent="primary"
          minimal
          style={{
            border: '1px solid var(--form-field-border)',
            width: '100%',
            display: 'flex',
            justifyContent: 'space-between'
          }}
          icon={val && val.icon && val.icon.name}
          rightIcon="chevron-down"
          text={val ? val.label : 'Search...'}
          withoutBoxShadow
        />
      </SelectV2>
    </Layout.Horizontal>
  )
}
export const AsyncSelect: Story<SelectV2Props> = args => {
  const { items = dummyPromise } = args
  const argsCopy = omit(args, ['items', 'itemRenderer', 'tagRenderer'])
  const [val, setVal] = useState<SelectOption>()
  return (
    <Layout.Horizontal id="primary-borderless-buttons">
      <SelectV2
        items={items}
        value={val}
        onChange={setVal}
        placeholder="xyz"
        itemRenderer={(item, props) => (
          <Button style={{ width: '100%', display: 'block' }} minimal onClick={ev => props.handleClick(ev as any)}>
            {item.label}
          </Button>
        )}
        {...argsCopy}>
        <Button
          intent="primary"
          minimal
          style={{ border: '1px solid var(--grey-400)', width: '100%', display: 'block' }}
          icon={val && val.icon && val.icon.name}
          rightIcon="chevron-down"
          text={val ? val.label : 'Search...'}
        />
      </SelectV2>
    </Layout.Horizontal>
  )
}
