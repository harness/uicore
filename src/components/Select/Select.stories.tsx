import React, { useState } from 'react'
import type { Meta, Story } from '@storybook/react'
import { Title, Subtitle, ArgsTable, Stories, PRIMARY_STORY, Primary, Description } from '@storybook/addon-docs/blocks'
import { Select, SelectProps } from '../..'
import { omit } from 'lodash-es'
import { SelectSize } from './Select'
import data from '../../_stories/components/pokedex.json'

export default {
  title: 'Form / Select',

  component: Select,
  parameters: {
    docs: {
      source: {
        type: 'code'
      },

      page: function PageDescription() {
        return (
          <>
            <Title>Select</Title>
            <Subtitle>
              <pre>
                <code>{`import {Select} from '@wings-software/uicore'`}</code>
              </pre>
            </Subtitle>
            <Description>{'SelectSize:small,medium,large'}</Description>
            <Primary />
            <ArgsTable story={PRIMARY_STORY} />

            <Stories />
          </>
        )
      }
    }
  },
  decorators: [Story => <Story />]
} as Meta
export const Basic: Story<SelectProps> = args => {
  const {
    items = [
      { label: 'Kubernetes', value: 'service-kubernetes' },
      { label: 'GitHub', value: 'service-github' },
      { label: 'ELK', value: 'service-elk' },
      { label: 'Jenkins', value: 'service-jenkins' },
      { label: 'GCP', value: 'service-gcp' }
    ]
  } = args
  const argsCopy = omit(args, ['size', 'items'])
  return <Select items={items} addClearBtn={true} {...argsCopy} />
}
Basic.args = { size: SelectSize.Large }
export const SelectWithIcons: Story<SelectProps> = args => {
  const {
    size = SelectSize.Large,
    items = [
      { label: 'Kubernetes', value: 'service-kubernetes', icon: { name: 'service-kubernetes' } },
      { label: 'GitHub', value: 'service-github', icon: { name: 'service-github' } },
      { label: 'ELK', value: 'service-elk', icon: { name: 'service-elk' } },
      { label: 'Jenkins', value: 'service-jenkins', icon: { name: 'service-jenkins' } },
      { label: 'GCP', value: 'service-gcp', icon: { name: 'service-gcp' } }
    ]
  } = args
  const argsCopy = omit(args, ['size', 'items'])
  return <Select size={size} items={items} addClearBtn={true} {...argsCopy} />
}
SelectWithIcons.args = { size: SelectSize.Large }
export const AsyncSelect: Story<SelectProps> = args => {
  interface SelectOption {
    label: string
    value: string | number | symbol
    icon?: any
  }

  const itemsLocal: SelectOption[] = data.map(row => ({
    label: row.name,
    value: row.id,
    icon: row.icon
  }))

  function dummyPromise(): Promise<SelectOption[]> {
    return new Promise<SelectOption[]>(resolve => {
      setTimeout(() => {
        resolve(itemsLocal)
      }, 5000)
    })
  }
  const [val, setVal] = useState<any>()
  const { items = dummyPromise } = args
  const argsCopy = omit(args, ['items'])
  return <Select items={items} value={val} onChange={val => setVal(val as any)} {...argsCopy} />
}
