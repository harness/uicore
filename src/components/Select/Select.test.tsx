import React from 'react'
import { render, fireEvent, act, wait } from '@testing-library/react'

import { Select, SelectOption } from './Select'

import data from '../../_stories/components/pokedex.json'

const items: SelectOption[] = data
  .map(row => ({
    label: row.name,
    value: row.id
  }))
  .slice(0, 20)

describe('<Select/> tests', () => {
  beforeAll(() => {
    jest.useFakeTimers()
  })

  test('works with static data', () => {
    const { container } = render(<Select items={items} />)

    expect(container).toMatchSnapshot()

    const input = container.querySelector('.bp3-input')

    expect(input).toBeDefined()
  })

  test('works with async data', async () => {
    const TIMEOUT = 5000
    const asyncData = () => {
      return new Promise<SelectOption[]>(resolve => {
        setTimeout(() => {
          resolve(items)
        }, TIMEOUT)
      })
    }
    const { container, queryByText } = render(<Select items={asyncData} />)
    expect(container).toMatchSnapshot()

    const input = container.querySelector('.bp3-input')!

    expect(input).toBeDefined()
    fireEvent.focus(input)

    expect(container).toMatchSnapshot('Loading state')
    expect(queryByText('Loading results...')).toBeDefined()

    await act(() => {
      jest.runTimersToTime(TIMEOUT)

      return wait()
    })

    expect(container).toMatchSnapshot('Final state')
    expect(container.querySelectorAll('li.menuItem').length).toBe(items.length)
  })

  test('basic filter works', () => {
    const { container } = render(<Select items={items} />)
    const input = container.querySelector('.bp3-input')!

    fireEvent.focus(input)

    expect(container).toMatchSnapshot('Initial Menu')
    expect(container.querySelectorAll('li.menuItem').length).toBe(items.length)

    fireEvent.change(input, { target: { value: 'cha' } })

    expect(container).toMatchSnapshot('Filtered Menu')
    expect(container.querySelectorAll('li.menuItem').length).toBe(3)
  })

  test('no matching results', () => {
    const { container, queryByText } = render(<Select items={items} />)
    const input = container.querySelector('.bp3-input')!

    fireEvent.focus(input)

    expect(container.querySelectorAll('li.menuItem').length).toBe(items.length)

    fireEvent.change(input, { target: { value: 'abcde' } })

    expect(container).toMatchSnapshot('Filtered Menu')
    expect(queryByText('No matching results found')).toBeDefined()
  })

  test('onChange', () => {
    const onChange = jest.fn()
    const { container, queryByText } = render(<Select items={items} onChange={onChange} />)
    const input = container.querySelector('.bp3-input')!

    fireEvent.focus(input)
    fireEvent.change(input, { target: { value: 'char' } })

    const pokemon = items.find(item => item.label === 'Charizard')
    const opt = queryByText(pokemon!.label)
    expect(opt).toBeDefined()

    if (opt) {
      fireEvent.click(opt)
      expect(onChange).toHaveBeenCalledWith(pokemon)
    }
  })
})
