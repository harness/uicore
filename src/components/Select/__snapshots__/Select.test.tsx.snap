// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select/> tests basic filter works: Filtered Menu 1`] = `
<div>
  <div
    class="bp3-popover-wrapper main bp3-fill"
  >
    <div
      class="bp3-popover-target bp3-popover-open"
    >
      <div
        class="bp3-input-group"
      >
        <input
          class="bp3-input"
          placeholder="Search..."
          style="padding-right: 0px;"
          type="text"
          value="cha"
        />
        <span
          class="bp3-input-action"
        >
          <span
            class="bp3-icon bp3-icon-chevron-down main padding padding-small"
            icon="chevron-down"
          >
            <svg
              data-icon="chevron-down"
              height="14"
              viewBox="0 0 16 16"
              width="14"
            >
              <desc>
                chevron-down
              </desc>
              <path
                d="M12 5c-.28 0-.53.11-.71.29L8 8.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42l4 4c.18.18.43.29.71.29s.53-.11.71-.29l4-4A1.003 1.003 0 0012 5z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </span>
      </div>
    </div>
    <div
      class="bp3-overlay bp3-overlay-open bp3-overlay-inline"
    >
      <div
        class="bp3-transition-container bp3-popover-appear bp3-popover-appear-active"
        style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none;"
      >
        <div
          class="bp3-popover bp3-minimal bp3-select-popover popover"
        >
          <div
            class="bp3-popover-content"
          >
            <div>
              <ul
                class="bp3-menu"
              >
                <li
                  class="menuItem active"
                >
                  Charmander
                </li>
                <li
                  class="menuItem"
                >
                  Charmeleon
                </li>
                <li
                  class="menuItem"
                >
                  Charizard
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Select/> tests basic filter works: Initial Menu 1`] = `
<div>
  <div
    class="bp3-popover-wrapper main bp3-fill"
  >
    <div
      class="bp3-popover-target bp3-popover-open"
    >
      <div
        class="bp3-input-group"
      >
        <input
          class="bp3-input"
          placeholder="Search..."
          style="padding-right: 0px;"
          type="text"
          value=""
        />
        <span
          class="bp3-input-action"
        >
          <span
            class="bp3-icon bp3-icon-chevron-down main padding padding-small"
            icon="chevron-down"
          >
            <svg
              data-icon="chevron-down"
              height="14"
              viewBox="0 0 16 16"
              width="14"
            >
              <desc>
                chevron-down
              </desc>
              <path
                d="M12 5c-.28 0-.53.11-.71.29L8 8.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42l4 4c.18.18.43.29.71.29s.53-.11.71-.29l4-4A1.003 1.003 0 0012 5z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </span>
      </div>
    </div>
    <div
      class="bp3-overlay bp3-overlay-open bp3-overlay-inline"
    >
      <div
        class="bp3-transition-container bp3-popover-appear bp3-popover-appear-active"
        style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none;"
      >
        <div
          class="bp3-popover bp3-minimal bp3-select-popover popover"
        >
          <div
            class="bp3-popover-content"
          >
            <div>
              <ul
                class="bp3-menu"
              >
                <li
                  class="menuItem active"
                >
                  Bulbasaur
                </li>
                <li
                  class="menuItem"
                >
                  Ivysaur
                </li>
                <li
                  class="menuItem"
                >
                  Venusaur
                </li>
                <li
                  class="menuItem"
                >
                  Charmander
                </li>
                <li
                  class="menuItem"
                >
                  Charmeleon
                </li>
                <li
                  class="menuItem"
                >
                  Charizard
                </li>
                <li
                  class="menuItem"
                >
                  Squirtle
                </li>
                <li
                  class="menuItem"
                >
                  Wartortle
                </li>
                <li
                  class="menuItem"
                >
                  Blastoise
                </li>
                <li
                  class="menuItem"
                >
                  Caterpie
                </li>
                <li
                  class="menuItem"
                >
                  Metapod
                </li>
                <li
                  class="menuItem"
                >
                  Butterfree
                </li>
                <li
                  class="menuItem"
                >
                  Weedle
                </li>
                <li
                  class="menuItem"
                >
                  Kakuna
                </li>
                <li
                  class="menuItem"
                >
                  Beedrill
                </li>
                <li
                  class="menuItem"
                >
                  Pidgey
                </li>
                <li
                  class="menuItem"
                >
                  Pidgeotto
                </li>
                <li
                  class="menuItem"
                >
                  Pidgeot
                </li>
                <li
                  class="menuItem"
                >
                  Rattata
                </li>
                <li
                  class="menuItem"
                >
                  Raticate
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Select/> tests no matching results: Filtered Menu 1`] = `
<div>
  <div
    class="bp3-popover-wrapper main bp3-fill"
  >
    <div
      class="bp3-popover-target bp3-popover-open"
    >
      <div
        class="bp3-input-group"
      >
        <input
          class="bp3-input"
          placeholder="Search..."
          style="padding-right: 0px;"
          type="text"
          value="abcde"
        />
        <span
          class="bp3-input-action"
        >
          <span
            class="bp3-icon bp3-icon-chevron-down main padding padding-small"
            icon="chevron-down"
          >
            <svg
              data-icon="chevron-down"
              height="14"
              viewBox="0 0 16 16"
              width="14"
            >
              <desc>
                chevron-down
              </desc>
              <path
                d="M12 5c-.28 0-.53.11-.71.29L8 8.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42l4 4c.18.18.43.29.71.29s.53-.11.71-.29l4-4A1.003 1.003 0 0012 5z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </span>
      </div>
    </div>
    <div
      class="bp3-overlay bp3-overlay-open bp3-overlay-inline"
    >
      <div
        class="bp3-transition-container bp3-popover-appear bp3-popover-appear-active"
        style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none;"
      >
        <div
          class="bp3-popover bp3-minimal bp3-select-popover popover"
        >
          <div
            class="bp3-popover-content"
          >
            <div />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Select/> tests works with async data 1`] = `
<div>
  <div
    class="bp3-popover-wrapper main bp3-fill"
  >
    <div
      class="bp3-popover-target"
    >
      <div
        class="bp3-input-group"
      >
        <input
          class="bp3-input"
          placeholder="Search..."
          style="padding-right: 0px;"
          type="text"
          value=""
        />
        <span
          class="bp3-input-action"
        >
          <span
            class="bp3-icon bp3-icon-chevron-down main padding padding-small"
            icon="chevron-down"
          >
            <svg
              data-icon="chevron-down"
              height="14"
              viewBox="0 0 16 16"
              width="14"
            >
              <desc>
                chevron-down
              </desc>
              <path
                d="M12 5c-.28 0-.53.11-.71.29L8 8.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42l4 4c.18.18.43.29.71.29s.53-.11.71-.29l4-4A1.003 1.003 0 0012 5z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Select/> tests works with async data: Loading state 1`] = `
<div>
  <div
    class="bp3-popover-wrapper main bp3-fill"
  >
    <div
      class="bp3-popover-target bp3-popover-open"
    >
      <div
        class="bp3-input-group"
      >
        <input
          class="bp3-input"
          placeholder="Search..."
          style="padding-right: 0px;"
          type="text"
          value=""
        />
        <span
          class="bp3-input-action"
        >
          <span
            class="bp3-icon bp3-icon-chevron-down main padding padding-small"
            icon="chevron-down"
          >
            <svg
              data-icon="chevron-down"
              height="14"
              viewBox="0 0 16 16"
              width="14"
            >
              <desc>
                chevron-down
              </desc>
              <path
                d="M12 5c-.28 0-.53.11-.71.29L8 8.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42l4 4c.18.18.43.29.71.29s.53-.11.71-.29l4-4A1.003 1.003 0 0012 5z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </span>
      </div>
    </div>
    <div
      class="bp3-overlay bp3-overlay-open bp3-overlay-inline"
    >
      <div
        class="bp3-transition-container bp3-popover-appear bp3-popover-appear-active"
        style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none;"
      >
        <div
          class="bp3-popover bp3-minimal bp3-select-popover popover"
        >
          <div
            class="bp3-popover-content"
          >
            <div>
              <ul
                class="bp3-menu"
              >
                <li
                  class="menuItem loading"
                >
                  Loading results...
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Select/> tests works with static data 1`] = `
<div>
  <div
    class="bp3-popover-wrapper main bp3-fill"
  >
    <div
      class="bp3-popover-target"
    >
      <div
        class="bp3-input-group"
      >
        <input
          class="bp3-input"
          placeholder="Search..."
          style="padding-right: 0px;"
          type="text"
          value=""
        />
        <span
          class="bp3-input-action"
        >
          <span
            class="bp3-icon bp3-icon-chevron-down main padding padding-small"
            icon="chevron-down"
          >
            <svg
              data-icon="chevron-down"
              height="14"
              viewBox="0 0 16 16"
              width="14"
            >
              <desc>
                chevron-down
              </desc>
              <path
                d="M12 5c-.28 0-.53.11-.71.29L8 8.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42l4 4c.18.18.43.29.71.29s.53-.11.71-.29l4-4A1.003 1.003 0 0012 5z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </span>
      </div>
    </div>
  </div>
</div>
`;
